algorithms/indexing/bravais_settings.py

```diff
@@ -0,0 +1,366 @@
+    """
+    p1_solution = all_solutions[-1]
+    assert p1_solution.setting_number == 1, p1_solution.setting_number
+    rmsd_p1 = p1_solution.rmsd
+
+    for solution in all_solutions:
+        solution.recommended = False
+        if solution["max_angular_difference"] < 0.5:
+            if (
+                solution.min_cc is None or solution.min_cc < 0.5
+            ) and solution.rmsd > 1.5 * rmsd_p1:
+                continue
+        elif (
+            solution.min_cc is None or solution.min_cc < 0.7
+        ) and solution.rmsd > 2.0 * rmsd_p1:
+            continue
+        elif solution.rmsd > 3 * rmsd_p1:
+            continue
+        solution.recommended = True
+
+
+def refine_subgroup(args):
+    assert len(args) == 4
+    params, subgroup, used_reflections, experiments = args
+
+    used_reflections = copy.deepcopy(used_reflections)
+    triclinic_miller = used_reflections["miller_index"]
+    cb_op = subgroup["cb_op_inp_best"]
+    higher_symmetry_miller = cb_op.apply(triclinic_miller)
+    used_reflections["miller_index"] = higher_symmetry_miller
+    unrefined_crystal = copy.deepcopy(subgroup.unrefined_crystal)
+    for expt in experiments:
+        expt.crystal = unrefined_crystal
+
+    subgroup.max_cc = None
+    subgroup.min_cc = None
+    subgroup.correlation_coefficients = []
+    subgroup.cc_nrefs = []
+
+    with LoggingContext("dials.algorithms.refinement", level=logging.ERROR):
+        try:
+            outlier_algorithm = params.refinement.reflections.outlier.algorithm
+            sel = used_reflections.get_flags(used_reflections.flags.used_in_refinement)
+            if sel.all_eq(False):
+                # Soft outlier rejection if no used_in_refinement flag is set
+                params.refinement.reflections.outlier.algorithm = "tukey"
+                iqr_multiplier = (
+                    params.refinement.reflections.outlier.tukey.iqr_multiplier
+                )
+                params.refinement.reflections.outlier.tukey.iqr_multiplier = (
+                    2 * iqr_multiplier
+                )
+                sel = ~sel
+            else:
+                # Remove reflections not previously used in refinement
+                params.refinement.reflections.outlier.algorithm = "null"
+            refinery, refined, outliers = refine(
+                params, used_reflections.select(sel), experiments
+            )
+            params.refinement.reflections.outlier.algorithm = outlier_algorithm
+            refinery, refined, outliers = refine(
+                params, used_reflections, refinery.get_experiments()
+            )
+        except RuntimeError as e:
+            if (
+                str(e) == "scitbx Error: g0 - astry*astry -astrz*astrz <= 0."
+                or str(e) == "scitbx Error: g1-bstrz*bstrz <= 0."
+            ):
+                subgroup.refined_experiments = None
+                subgroup.rmsd = None
+                subgroup.Nmatches = None
+            else:
```

Comment:
wondering if bailing out here with `return subgroup` is clearer than `try: except: else`